### PR TITLE
lint: Fix lint-whitespace issues

### PR DESCRIPTION
**Pull request by maflcko**
The lint check has many issues:

* It uses `COMMIT_RANGE`, which is brittle code, apparently making it harder to run the CI locally, or self-hosted. See https://github.com/bitcoin/bitcoin/pull/29274#discussion_r1457739319 
* The result depends on `COMMIT_RANGE`, or the number of commits passed to the script, which can cause false negatives or false positives.
* It is based on the diff output, parsing it, and printing it again, which is brittle as well.
* The output does not include line number, making it harder to act on a lint error.

Fix all issues by removing the script and replacing it with a simple call to `git grep -I --line-number ...`.

### CODE REVIEW

**Comment by Sjors** [comment]
Does this PR allow dropping `COMMIT_RANGE` from `ci/lint/06_script.sh`?

**Comment by Empact** [comment]
Concept ACK

File: test/lint/test_runner/src/main.rs [hunk 1]
```diff
@@ -95,6 +95,85 @@ fs:: namespace, which has unsafe filesystem functions marked as deleted.
     }
 }
 
+/// Return the pathspecs for whitespace related excludes
+fn get_pathspecs_exclude_whitespace() -> Vec<String> {
+    let mut list = get_pathspecs_exclude_subtrees();
+    list.extend(
+        [
+            // Permanant excludes
```
Comment (flack) [inline test/lint/test_runner/src/main.rs:103]: Permanant => Permanent

File: test/lint/test_runner/src/main.rs [hunk 2]
```diff
@@ -95,6 +95,85 @@ fs:: namespace, which has unsafe filesystem functions marked as deleted.
...
+
+fn lint_trailing_whitespace() -> LintResult {
+    let trailing_space = git()
+        .args(["grep", "-I", "--line-number", "\\s$", "--"])
+        .args(get_pathspecs_exclude_whitespace())
+        .status()
+        .expect("command error")
+        .success();
+    if trailing_space {
+        Err(r#"
```
Comment (TheCharlatan) [inline test/lint/test_runner/src/main.rs:136]: How about displaying the offending lines with something like:
```diff
diff --git a/test/lint/test_runner/src/main.rs b/test/lint/test_runner/src/main.rs
index cffb477950..6f16620605 100644
--- a/test/lint/test_runner/src/main.rs
+++ b/test/lint/test_runner/src/main.rs
@@ -8 +8 @@ use std::path::{Path, PathBuf};
-use std::process::{Command, Stdio};
+use std::process::Command;
@@ -129 +129 @@ fn lint_trailing_whitespace() -> LintResult {
-    let output = git()
+    let output = check_output(git()
@@ -131,6 +131,4 @@ fn lint_trailing_whitespace() -> LintResult {
-        .args(get_pathspecs_exclude_whitespace())
-        .output()
-        .expect("command error");
-    let stderr = String::from_utf8_lossy(&output.stderr);
-    if !stderr.is_empty() {
-        Err(r#"
+        .args(get_pathspecs_exclude_whitespace()));
+    if let Ok(message) = output {
+        Err(format!(r#"
+{}
@@ -146,2 +144,2 @@ sourced files to the exclude list.
-            "#
-        .to_string())
+            "#,
+        message))
@@ -150 +148 @@ sourced files to the exclude list.
-    }    
+    }
@@ -154 +152 @@ fn lint_tabs_whitespace() -> LintResult {
-    let tabs = git()
+    let output= check_output(git()
@@ -157,6 +155,5 @@ fn lint_tabs_whitespace() -> LintResult {
-        .args(get_pathspecs_exclude_whitespace())
-        .status()
-        .expect("command error")
-        .success();
-    if tabs {
-        Err(r#"
+        .args(get_pathspecs_exclude_whitespace()));
+
+    if let Ok(message) = output {
+        Err(format!(r#"
+{}
@@ -170,2 +167,2 @@ Please add any false positives, such as subtrees, or externally sourced files to
-            "#
-        .to_string())
+            "#,
+        message))
```

File: test/lint/test_runner/src/main.rs [hunk 3]
```diff
@@ -95,6 +95,85 @@ fs:: namespace, which has unsafe filesystem functions marked as deleted.
...
+        .to_string())
+    } else {
+        Ok(())
+    }
+}
+
+fn lint_tabs_whitespace() -> LintResult {
+    let tabs = git()
+        .args(["grep", "-I", "--line-number", "--perl-regexp", "^\\t", "--"])
+        .args(["*.cpp", "*.h", "*.md", "*.py", "*.sh"])
```
Comment (TheCharlatan) [inline test/lint/test_runner/src/main.rs:156]: This list is pretty limited. Would adding .rs to it make sense?

File: test/lint/test_runner/src/main.rs (hunk 2 above)
Comment (TheCharlatan) [reply test/lint/test_runner/src/main.rs:136]: Ah, this is git paging the output on my system if the output is not piped to somewhere else. I'll just switch the paging off for `git grep` then, it's annoying anyway.

**Review by sedited — APPROVED** [review]
Nice, ACK fa5729436ca12b20cfa2cd1f0c6f54af7192f0a6

File: test/lint/test_runner/src/main.rs (hunk 2 above)
Comment (TheCharlatan) [reply test/lint/test_runner/src/main.rs:136]: Or just `git --no-pager`?

**Review by sedited — APPROVED** [review]
Re-ACK 5555395c15e896230a55c131fc3cbfd9d116adf8

**Pull request merged**